### PR TITLE
SI-6634 Fixes data corruption issue in ListBuffer#remove

### PR DESCRIPTION
**Pull request by soc**
Bounds are now enforced (and exceptions get thrown),
instead of trying to make invalid bounds work.

### CODE REVIEW

**Comment by soc** [comment]
See additional details in the bug tracker.

**Comment by jsuereth** [comment]
Review by@jsuereth and @axel22

File: src/library/scala/collection/mutable/ListBuffer.scala [hunk 1]
```diff
@@ -250,12 +250,15 @@ final class ListBuffer[A]
    *  @param count     the number of elements to remove.
    */
   override def remove(n: Int, count: Int) {
+    // We check the bounds early, so that we don't trigger copying.
+    if (n < 0 || n >= len) throw new IndexOutOfBoundsException(n.toString)
+    if (count < 0) throw new IndexOutOfBoundsException(count.toString)
```
Comment (axel22) [inline src/library/scala/collection/mutable/ListBuffer.scala]: Should the last one be an `IllegalArgumentException`?

**Comment by axel22** [comment]
I'd consider merging the two test files into one in the interest of keeping the test suite smaller (and faster).
Also, should one of the exceptions thrown be an `IllegalArgumentException`?

Otherwise, LGTM.

**Comment by soc** [comment]
@axel22 I'll read up on it, as far as I know there is a fairly precise description in the Java world for that.
I'm thinking about revising the exception messages of this and the other pull request anyway ...

**Comment by soc** [comment]
Review by @jsuereth @axel22 
- Changed one `IOOBE` to an `IAE` as specified in the interface
- Added more tests
- Added more descriptive exception messages

A general note:

I think we really need to go through the standard library and fix our exception messages. I don't believe there is any reason why our exception messages have to suck as much as Java's.

Additionally, it might make sense to factor out common exception checks/messages and move them to a supertype, instead of duplicating slightly different messages all over the place. E. g. move all buffer related exceptions to `Buffer`'s or `BufferLike`s companion object.

**Comment by axel22** [comment]
LGTM.

**Comment by soc** [comment]
Mhhh. I think the last bounds check should also be an `IAE` instead of an `IOOBE`, because the index is valid, but not the count (like in the other case where we throw an `IAE`).

**Comment by soc** [comment]
Any opinions on that? This would also lead to a small change/clarification in `BufferLike` and possibly other implementation.

Normally, I'd be wary of such a change, but considering that code causing this exceptions is very likely wrong anyway and that it would simplify exception handling for the user (`IOOBE` => index is invalid and `IAE` => count is invalid, not some inconsistent mixture) I'd like to make this change. (If `@migration` gets resurrected, I'd also add a notice to it.)

Opinions?

**Comment by paulp** [comment]
If it isn't going into 2.10.0, it probably shouldn't go into 2.10.x, but other than that I'm all in favor of it (especially in favor of using named exceptions in a rigorously consistent fashion.)

**Comment by soc** [comment]
I think fixes for the corruption issues could go into 2.10.1. The current PR should probably be appropriate as is, so I won't add anything on top of it.

If this is OK, I'll target 2.11 for further improvements like consistent exception types, helpful exception messages, better parameter names, etc...

**Comment by paulp** [comment]
Right, sorry, I only meant don't throw a different exception in 2.10.x than is thrown presently.

**Comment by soc** [comment]
No problem, but could you clarify what exactly should happen for 2.10.x?

```
     Error case      | What ScalaDoc says |   What ListBuffer does   | What ArrayBuffer does | What should happen (imho)
-------------------------------------------------------------------------------------------------------------------------
 (1) index negative  |      IOOBE         |– (Removes element from 0)|      IOOBE            |      IOOBE
 (2) index too large |      IOOBE         |– (DS corrupted, UOE/NSEE)|      IOOBE            |      IOOBE
 (3) count negative  |      IAE           |– (DS corrupted, NSEE)    |      IAE              |      IAE
 (4) count too large |      IOOBE         |– (Removes elems 'til end)|      IOOBE            |      IAE
```

Only fix `(3)`?

**Comment by paulp** [comment]
Someone who isn't me should really determine this.

**Comment by axel22** [comment]
In my opinion:
- a `count` can only be non-negative - a negative count is illegal, so it should throw an `IAE`
- any combinations of arguments that results in indexing out of bounds of the buffer should throw an `IOOBE`

Examples of the latter:
- `index` outside valid buffer index range - `IOOBE`
- `index + count` outside valid buffer index range - `IOOBE`

So, to be consistent with negative count not meaning anything, indexing out of bounds meaning an index out of bounds exception and to be consistent with `ArrayBuffer`, I'd say (4) needs to be `IOOBE`.
That, I think (1)-(4) should behave exactly like `ArrayBuffer`.

**Comment by soc** [comment]
Mhhh, OK. Just asked another person familiar with Java and common practices and he told me (4) should be chosen...

Personally, I'd also prefer if we had a clear separation where `index` only throws `IOOBE` and `count` only throws `IAE`. Reason: what the implementation does underneath with `index` and `count` is an implementation detail and should not be the concern of the user. I think it is inconsistent to have `count` throw different exceptions depending on the value.

TL;DR: I would prefer this for consistency reasons and because I think the rules are more clear and memorable for users:
`IOOBE` => index argument wrong
`IAE` => count argument wrong

To make (at least this PR) short:

I'd suggest that we should merge the fairly uncontroversial fix for (3) in #1597 and the whole #1596 for 2.10.x, the rest can be discussed in an upcoming PR against 2.11. (I will write a larger email proposing a thorough cleanup containing consistent exception types, better exception messages and consistent parameter names for 2.11.)

**Comment by soc** [comment]
Discovered yet another error case with yet another exception type ... I'm still working on a minimally invasive fix, which doesn't change exception types, keeps the current bounds behaviour and prevents corruption of the data structure.

**Comment by soc** [comment]
I updated the table above to indicate which exceptions are not thrown immediately, but later due to corrupted data structures.
I pushed a new fix, with the following properties:

```
     Error case      | What ScalaDoc says |   What ListBuffer does   | What ArrayBuffer does | What should happen (imho)
-------------------------------------------------------------------------------------------------------------------------
 (1) index negative  |      IOOBE         |– (Removes element from 0)|      IOOBE            |      IOOBE
 (2) index too large |      IOOBE         |– (Does nothing)          |      IOOBE            |      IOOBE
 (3) count negative  |      IAE           |          IAE             |      IAE              |      IAE
 (4) count too large |      IOOBE         |– (Removes elems 'til end)|      IOOBE            |      IAE
```

I tried to be as lenient as possible and only fix the data structure corruption, but I can also make it more strict and reject more invalid inputs.

**Comment by jsuereth** [comment]
OK, as is, this LGTM.   We should discuss collection design in a bit more depth, perhaps start an ML thread or some other forum for discussion design that's slightly better than PR comments.
